M/MKL/build_tarballs.jl

```diff
@@ -3,38 +3,70 @@ using BinaryBuilder, Pkg
 name = "MKL"
 version = v"2021.1.1"
 
-sources = [
-    ArchiveSource("https://anaconda.org/intel/mkl/2021.1.1/download/linux-64/mkl-2021.1.1-intel_52.tar.bz2",
-                  "bfb0fd056576cad99ae1d9c69ada2745420da9f9cf052551d5b91f797538bda2"; unpack_target = "mkl-x86_64-linux-gnu"),
-    ArchiveSource("https://anaconda.org/intel/mkl/2021.1.1/download/linux-32/mkl-2021.1.1-intel_52.tar.bz2",
-                  "7b6f55a30886154bd96d4b4c6b7428494a59397b87779b58e5b3de00250343f9"; unpack_target = "mkl-i686-linux-gnu"),
-    ArchiveSource("https://anaconda.org/intel/mkl/2021.1.1/download/osx-64/mkl-2021.1.1-intel_50.tar.bz2",
-                  "819fb8875909d4d024e2a936c54b561aebd1e3aebe58fc605c70aa1ad9a66b70"; unpack_target = "mkl-x86_64-apple-darwin14"),
-    ArchiveSource("https://anaconda.org/intel/mkl/2021.1.1/download/win-32/mkl-2021.1.1-intel_52.tar.bz2",
-                  "dba6a12a481407ec55fba9895b68afacb15f044905dcb5e185db341b688e6177"; unpack_target = "mkl-i686-w64-mingw32"),
-    ArchiveSource("https://anaconda.org/intel/mkl/2021.1.1/download/win-64/mkl-2021.1.1-intel_52.tar.bz2",
-                  "4024391b8a45836d5a7ee92405b7767874b3c3bbf2f490349fda042db3b60dfd"; unpack_target = "mkl-x86_64-w64-mingw32"),
-]
-
-# Bash recipe for building across all platforms
-script = raw"""
-cd ${WORKSPACE}/srcdir/mkl-${target}
-if [[ ${target} == *-mingw* ]]; then
-    cp -r Library/bin/* ${libdir}
-else
-    cp -r lib/* ${libdir}
-fi
+# Bash recipes for building across all platforms
+script = read(joinpath(@__DIR__, "script.sh"), String)
+script_macos = read(joinpath(@__DIR__, "script_macos.sh"), String)
 
-install_license info/licenses/*.txt
-"""
+non_reg_ARGS = filter(arg -> arg != "--register", ARGS)
 
-platforms = [
-    Platform("x86_64", "linux"; libc="glibc"),
-    Platform("i686", "linux"; libc="glibc"),
-    Platform("x86_64", "macos"),
-    Platform("i686", "windows"),
-    Platform("x86_64", "windows"),
-]
+platform_sources = Dict(
+    Platform("x86_64", "linux"; libc="glibc") => (
+        source = ArchiveSource(
+            "https://anaconda.org/intel/mkl/2021.1.1/download/linux-64/mkl-2021.1.1-intel_52.tar.bz2",
+            "bfb0fd056576cad99ae1d9c69ada2745420da9f9cf052551d5b91f797538bda2";
+            unpack_target = "mkl-x86_64-linux-gnu"
+        ),
+        # We need to run autofix on Linux, because here libmkl_rt doesn't
+        # have a soname, so we can't ccall it without specifying the path:
+        # https://github.com/JuliaSparse/Pardiso.jl/issues/69
+        autofix = true,
+        script = script,
+        args = ARGS,
```

Comment:
The whole difference between `non_reg_args` and `ARGS` is actually a little subtle: we want to disable `--register` on the first invocations not because of the platforms, but because of the ordering.  So if we loop over the platform sources in dictionary order, we aren't guaranteed that the `--register` will be applied to the last invocation.  :(

Reply:
Okay, thanks--I was a bit confused by this.  I'll change the `Dict` to simply be a list of `NamedTuples`, with `platform` as the first member (there's no real need for it to be a `Dict`), remove `args`, and make sure we use `non_reg_args` on the first set of invocations, and `ARGS` on the last.

Reply:
Fixed (hopefully) in latest commit.  If this looks okay, I'll update the first commit.